src/fastify-plugins/maps/offline-fallback-map.js

```diff
@@ -0,0 +1,117 @@
+import path from 'path'
+import fs from 'fs/promises'
+import FastifyStatic from '@fastify/static'
+import fp from 'fastify-plugin'
+
+import {
+  NotFoundError,
+  createStyleJsonResponseHeaders,
+  getFastifyServerAddress,
+} from '../utils.js'
+
+export const PLUGIN_NAME = 'mapeo-static-maps'
+
+export const plugin = fp(offlineFallbackMapPlugin, {
+  fastify: '4.x',
+  name: PLUGIN_NAME,
+})
+
+/**
+ * @typedef {object} OfflineFallbackMapPluginOpts
+ * @property {string} [prefix]
+ * @property {string} styleJsonPath
+ * @property {string} sourcesDir
+ */
+
+/**
+ * @typedef {object} FallbackMapPluginDecorator
+ * @property {(serverAddress: string) => Promise<any>} getResolvedStyleJson
+ * @property {() => Promise<import('node:fs').Stats>} getStyleJsonStats
+ */
+
+/** @type {import('fastify').FastifyPluginAsync<OfflineFallbackMapPluginOpts>} */
+async function offlineFallbackMapPlugin(fastify, opts) {
+  const { styleJsonPath, sourcesDir } = opts
+
+  fastify.decorate(
+    'mapeoFallbackMap',
+    /** @type {FallbackMapPluginDecorator} */
+    ({
+      async getResolvedStyleJson(serverAddress) {
+        const rawStyleJson = await fs.readFile(styleJsonPath, 'utf-8')
+        const styleJson = JSON.parse(rawStyleJson)
+
+        const sources = styleJson.sources || {}
+
+        const sourcesDirFiles = await fs.readdir(sourcesDir, {
+          withFileTypes: true,
+        })
+
+        for (const file of sourcesDirFiles) {
+          // Only work with files
+          if (file.isDirectory()) continue
+
+          // Ignore the style.json file if it exists
+          if (file.name === 'style.json') continue
+
+          // Only work with json or geojson files
+          const extension = path.extname(file.name)
+          if (!(extension === '.json' || extension === '.geojson')) continue
+
+          const sourceName = path.basename(file.name, extension) + '-source'
+
+          sources[sourceName] = {
+            type: 'geojson',
+            data: new URL(`${opts.prefix || ''}/${file.name}`, serverAddress)
+              .href,
+          }
+        }
+
+        styleJson.sources = sources
+
+        return styleJson
+      },
+      async getStyleJsonStats() {
```

Comment:
Nit: because this returns a promise, we don't need to mark it `async`.

```suggestion
      getStyleJsonStats() {
```

Reply:
addressed via https://github.com/digidem/mapeo-core-next/pull/450/commits/da7b7b9c47a4263a7befdd07d3f8ea57cd04cfd6